### DISCUSSION TITLE
Switch layout based on screen index

### DISCUSSION THREAD

**Discussion by GustawXYZ** [qtile/qtile#4350] [General]
Hello, I'm trying to have layouts assigned to screens. My idea as of now was to simply change the layout of given group, but I can't figure it out.

What I have now is (config without following changes for refference: https://gitlab.com/centeepede/dotfiles/-/blob/main/qtile/functions.py):
```python
file:functions.py

class Functions:
    @staticmethod
    def switch_group_and_change_layout(group_name):
        @lazy.function
        def __inner(qtile):
            from libqtile.log_utils import logger
            logger.warning(f"Switching to group: {group_name}")
            lazy.group[group_name].toscreen()
            i = qtile.screens.index(qtile.current_screen)
            if i == 0:
                ##  Edit based on the first comment:
                # - lazy.group[group_name].layout("monadtall")
                lazy.group[group_name].setlayout("monadtall")
            elif i > 0:
                ## Edit based on the first comment:
                # - lazy.group[group_name].layout("monadwide")
                lazy.group[group_name].setlayout("monadwide")
        return __inner
```
```python
file:keys.py
 
from functions import Functions
...
           # old call that worked, but didn't change the layout
            # group_keys += [Key([MOVEMENT_KEY], index, lazy.group[icon].toscreen()), Key(
            group_keys += [Key([MOD], index, Functions.switch_group_and_change_layout(icon)), Key(
                [MOD, SHIFT], index, lazy.window.togroup(icon, switch_group=False))]
...
```
But I have two problems, first: groups don't get changed and there are no errors if I comment out everything in switch_group_and_change_layout like so:
```python
file:functions.py

class Functions:
    @staticmethod
    def switch_group_and_change_layout(group_name):
        @lazy.function
        def __inner(qtile):
            lazy.group[group_name].toscreen()
        return __inner
```
It looks to get called, but behaves like a noop.

Second issue is I'm not sure how to set the layout, because my `lazy.group[group_name].layout("monadtall")`
results in following error:
```
and_change_layout.<locals>.__inner at 0x7f424cb78f40>":
Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/libqtile/command/base.py", line 213, in cmd_function
    function(self, *args, **kwargs)
  File "/home/gustaw/.config/qtile/functions.py", line 41, in __inner
    lazy.group[group_name].layout("monadtall")
  File "/usr/lib/python3.11/site-packages/libqtile/command/client.py", line 197, in __call__
    raise SelectError("Invalid call", "", self._current_node.selectors)
libqtile.command.base.SelectError: Invalid call, name: , selectors: [('group', '2 \uf120 '), ('layout', None)]
```
I would appreciate any help here :D

It looks that layouts are persistent once I set them manually, but setting them on initially by hand is annoying, so maybe there's a better way to configure Screens to have preferred layout I'm missing?

**Comment by ervinpopescu** [comment]
I don't think `_Group` exposes `layout` command, you need to use `setlayout`, with a `Layout` object. See [here](https://docs.qtile.org/en/latest/manual/commands/api/groups.html#libqtile.group._Group). I think you can use the layouts you already declare in the config and pass the wanted layout to `setlayout`, as that would pass it by reference and `qtile` would know how to deal with it. Not an expert tho, so don't quote me on it.

**Comment by GustawXYZ** [comment]
Thanks a lot! Missed the @expose_command() decorators!
It looks to be working with string name:
```
lazy.group[group_name].setlayout(layouts[0].name)
# or
lazy.group[group_name].setlayout("monadtall")
```

I still however get no action at all from lazy calls that are inside of @lazy.functions. Both group change and set layout, seem do nothing...
But I do get the logs indicating that this was called with what looks like correct values:
```
2023-07-23 16:53:51,667 WARNING libqtile keys.py:switch_groups():L230 Switching to group: 2 
2023-07-23 16:53:52,062 WARNING libqtile keys.py:switch_groups():L230 Switching to group: 3 
2023-07-23 16:53:52,385 WARNING libqtile keys.py:switch_groups():L230 Switching to group: 4 
``` 
Do you have any ideas how to deal with this?

**Reply by elParaguayo** [reply]
You can't call a lazy function from within another.  Lazy commands do not actually run the command; they create a reference to a command which the server can run when the command is passed to it. You need to use the internal qtile commands inside a function. 

Which commands are you trying to run?

**Comment by GustawXYZ** [comment]
Thanks! I just found you post on reddit: https://www.reddit.com/r/qtile/comments/yc6cpu/how_to_run_any_lazy_function_in_custom_function/ 
which made me realize that I was looking at the implementation from git, whilst being on 0.22.1! So basically I had to add `cmd_` to switch from `lazy....` to `qtile....`...

Alright, it looks like it works now with this:

```python
    def init_keys_groups(self, group_names):
        """
        Create bindings to move between groups
        """
        @lazy.function
        def switch_groups(qtile, group_name):
            current_screen_index = qtile.screens.index(qtile.current_screen)
            layout_index = "monadtall"
            if current_screen_index == 0:
                layout_index = "monadtall"
            elif current_screen_index > 0:
                layout_index = "monadwide"
            for g in qtile.groups:
                if g.name == group_name:
                    # change to g.tooscreen() when it's released from git
                    g.cmd_toscreen()
                    # change to qtile.cmd_to_layout_index when it's released from git
                    g.cmd_setlayout(layout_index)
                    return

        group_keys = []
        for group_name in group_names:
            index = (group_name[0]).lower()
            group_keys += [Key([MOD], index, switch_groups(group_name)), Key(
                [MOD, SHIFT], index, lazy.window.togroup(group_name, switch_group=False))]

        return group_keys        

```

So that solves this mostly. 

I guess one TODO left is that top bar buttons don't implement this, so probably I should move this functionality to a hook (somehow) instead of relying on keys.
Or maybe @elParaguayo do you have other more elegant suggestions?
